contents/projectManagement/testDrivenDevelopment.md

```diff
@@ -0,0 +1,199 @@
+
+The tests also cover all possible branches and paths that the production code can take because no more code is written than is necessary[^3]. Detecting bugs at early stages is easier when you're working with code that has high coverage [^5]. 
+
+<modal title="Mock Objects and Frameworks" id="modal:index-mock">
+Mock objects are simulated objects whose sole purpose is to act as a black-box that mimics the behaviour of real-world objects during testing. They are especially useful in writing unit tests, which are meant to test the functionality of a portion of code assuming that the rest of the code behaves as it should. 
+</modal>
+
+### Benefit 2: Ensures that Product Requirements are Understood
+
+Test-driven development requires the developers to have a good understanding of the product or feature requirements before any sort of development can begin. As mentioned earlier, the first step in development is (ideally) writing tests. Coming up with tests without a clear picture of the requirements or specifications  can be dangerous as - 
+
+1. you could end up testing something that doesn't need to be tested, thus giving you a false sense of accomplishment when you write production code to pass the test, or 
+1. you could end up overlooking some requirements that _must_ be tested, which is far worse.
+
+TDD thus prevents the team from rushing into development, and pushes them to prioritize the requirements-gathering step. 
+
+
+
+### Benefit 3: Increases Productivity
+
+Hakan Erdogmus, Professor of Software Engineering at Carnegie Mellon University, found that programmers who wrote more tests tended to be more productive^[Erdogmus, 2005. ["On the Effectiveness of Test-first Approach to Programming"](https://web.archive.org/web/20141222180731/http://nparc.cisti-icist.nrc-cnrc.gc.ca/npsi/ctrl?action=shwart&index=an&req=5763742&lang=en)], and since employing TDD meant writing more tests, there is a correlation between productivity and TDD.
+
+Employing TDD also lays out the design of a program. It forces programmers to think about coming up with failing tests and working their way up to writing code that passes the test. Furthermore, since tests are written first, there is close to 100% coverage. This ensures that production code is covered by atleast one test. This can boost individial and team morale, confidence in the code, and productivity as a consequence.
+
+
+## Caveats and Limitations
+
+Even with some of the attractive benefits of adopting TDD, there is no one-size-fits-all solution for project design management. Some of the limitations of TDD are: 
+
+1. **Strict emphasis on communication of requirements and specifications**
```

Comment:
The reason I stuck with points instead of headers is to maintain consistency across book chapters. Maybe we can decide on a standard?

Reply:
Might be good to standardise with ### headings, since most of the articles on the site use that more often.

But its a really minor thing, I wouldn't mind it as a reader either way. Using points has the effect of giving slightly less emphasis to the caveats too, which is nice

Reply:
> Using points has the effect of giving slightly less emphasis to the caveats too, which is nice

That's true